src/core/Form/RadioButton/RadioButton.tsx

```diff
@@ -1,74 +1,193 @@
-    <DefaultRadioButton {...withSuomifiDefaultProps(props)} variant="large" />
-  );
+    const { checkedState } = this.state;
+    const hintTextId = `${id}-hintText`;
 
-  static divider = (props: RadioButtonDividerProps) => (
-    <RadioButtonDivider {...withSuomifiDefaultProps(props)} />
-  );
+    const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+      if (checked === undefined) {
+        this.setState({ checkedState: event.target.value });
+      }
+      if (!!onChange) {
+        onChange(event);
+      }
+    };
+
+    return (
+      <HtmlDiv
+        className={classnames(
+          baseClassName,
+          radioButtonClassNames.container,
+          className,
+          {
+            [radioButtonClassNames.disabled]: disabled,
+            [radioButtonClassNames.large]: variant === 'large',
+            [radioButtonClassNames.checked]: checked,
+          },
+        )}
+      >
+        <HtmlInput
+          className={radioButtonClassNames.input}
+          type="radio"
+          name={name}
+          id={id}
+          disabled={disabled}
+          onChange={handleChange}
+          checked={checkedState}
+          {...getConditionalAriaProp('aria-describedby', [
+            hintText ? hintTextId : undefined,
+          ])}
+          {...(value ? { value } : {})}
+          {...passProps}
+        />
+        <HtmlLabel htmlFor={id} className={radioButtonClassNames.label}>
+          {children}
+        </HtmlLabel>
+        {hintText && (
+          <HtmlSpan className={radioButtonClassNames.hintText} id={hintTextId}>
+            {hintText}
+          </HtmlSpan>
+        )}
+      </HtmlDiv>
+    );
+  }
+}
 
+const StyledRadioButton = styled(
+  ({ tokens, ...passProps }: RadioButtonProps & InternalTokensProp) => (
+    <BaseRadioButton {...passProps} />
+  ),
+)`
+  ${(tokens) => baseStyles(tokens)}
+`;
+
+export class RadioButton extends Component<RadioButtonProps> {
   render() {
-    return <DefaultRadioButton {...withSuomifiDefaultProps(this.props)} />;
+    const { id: propId, onChange, ...passProps } = this.props;
+    return (
+      <AutoId id={propId}>
+        {(id) => (
+          <RadioButtonGroupConsumer>
+            {(consumer) => (
+              <StyledRadioButton
+                {...withSuomifiDefaultProps(passProps)}
+                {...(!!consumer.selectedValue
+                  ? { checked: consumer.selectedValue === passProps.value }
+                  : {})}
+                {...(!!consumer.name ? { name: consumer.name } : {})}
+                id={id}
+                onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
+                  if (!!consumer.onRadioButtonChange) {
+                    consumer.onRadioButtonChange(event.target.value);
+                  }
+                  if (!!onChange) {
+                    onChange(event);
+                  }
+                }}
+              />
+            )}
```

Comment:
Props from `consumer` could be also deconstucted here, but I think the current way is clearer what goes where.
I think this can be leaved as it is.

Reply:
I don't like this part so much, but could not find a better way without polluting the underlying components with external state / consumer.

Reply:
It's now using destructuring and a bit cleaner in my opinion. Not clean by any means, but would probably require additional tooling to make it better.

Reply:
Just tested that we might be able to shorten even more,
```tsx
{...(!!onRadioButtonChange
  ? {
      checked: selectedValue === passProps.value,
      name,
    }
  : {})}
```
to be like this:
```tsx
{...(!!onRadioButtonChange && {
  checked: selectedValue === passProps.value,
  name,
})}
```

Didn't test it very thoroughly yet, but what do you think about this?

Reply:
It might work, but not sure what this would transpile into. We would be more or less spreading undefined but not inside of an object. Needs further investigation. https://stackoverflow.com/questions/46957194/javascript-es6-spread-operator-on-undefined

Reply:
Maybe not worth the risk to change this.